lib/gameplay.rb

```diff
@@ -144,29 +147,64 @@ def display_boards
   def user_turn
     puts "Please select a coordinate to fire upon. Choose wisely."
       shot = gets.chomp
+      if @computer_board.cells[shot].fired_upon? == false
         if @computer_board.valid_coordinate?(shot)
         @computer_board.cells[shot].fire_upon
         sleep(1.5)
-        puts "Your shot on #{shot} was a #{@computer_board.cells[shot].render}"
+        explain_render(@computer_board, shot, "your")
         sleep(1.5)
+        else
+          puts "Invalid coordinates. Try again."
+          user_turn
         end
+      else
+        puts "This coordinate has already been fired upon.
+        Please choose another."
+        user_turn
+      end
       display_boards
       computer_turn
   end
 
   def computer_turn
     puts "I will now attempt to fire on one of your ships."
-    comp_shot = @player_board.cells.keys.sample(1).strip
-    if @player_board.valid_coordinate?(comp_shot)
-      @player_board.cells[comp_shot].fire_upon
+    shot = @player_board.cells.keys.sample
+    until @computer_board.cells[shot].fired_upon? == false do
+      shot = @player_board.cells.keys.sample
+    end
+
+    if @player_board.valid_coordinate?(shot)
+      @player_board.cells[shot].fire_upon
       sleep(1.5)
-      puts "My shot on #{comp_shot} was a #{@player_board.cells[comp_shot].render}"
+      explain_render(@player_board, shot, "my")
       sleep(1.5)
     end
     display_boards
     user_turn
   end
 
+  def explain_render(board, shot, pov)
+    if board.cells[shot].render == "M"
+      puts "#{pov.capitalize} shot on #{shot} was a miss."
+    elsif board.cells[shot].render == "H"
+      puts "#{pov.capitalize} shot on #{shot} was a HIT!"
```

Comment:
I woke up this morning so bugged that it was only putting "M" instead of miss, and look at you reading my mind :') Nice!